### ISSUE TITLE
Line 126 in AEI_Net.py

### ISSUE THREAD

**Issue by ocastan** [taotaonice/FaceShifter#16]
I'm still struggling with AEI learning: I had to reduce the batch size to 5 due to my card memory, but it collapse between 10 and 14 epochs. I'm trying another optimiser (ranger) which hasn't collapsed so far, but it's painfully slow.
I may try other tricks like different optimisers for G and D or gradient accumulation...
Currently my question is: why have you commented out torch,no_grad in get_attr function ? I can't understand why Y should contribute to gradient calculation.

**Comment by taotaonice** [comment]
You are right. At first, I didn't think about it very carefully, and recently I notice this. At the moment, I am training for that whether Y should contribute to gradient calculation, and now I think Y should be .detach().

**Comment by ocastan** [comment]
I'm beginning with ML but my understanding is that detach will make Y a leaf but that gradient will still accumulate in your encoder graph and on the other hand no_grad is supposed to disable any gradient accumulation. I would rather uncomment the line or do both

**Comment by taotaonice** [comment]
Your method is almost same with mine, they have the same objective of optimization.

**Comment by ocastan** [comment]
Ok. no_grad may be able to save more memory?

**Comment by ocastan** [comment]
After thinking about it, I'm not sure you should change anything: the attribute loss is the L2 distance between Yatt which is encoder(generator(encoder(Xt), z_id)) and Xatt which is encoder(Xt)
If you detach Y, the backprop will update the encoder weights, if you do not it will update both the encoder and the generator weights
But I'm not sure

**Comment by taotaonice** [comment]
I think L_attr should be just related with attr encoder, unrelated with generator, so I should only update weights of attr encoder.

**Comment by ocastan** [comment]
I give it a try since I have no attribute transfer after 38 epochs

**Comment by ocastan** [comment]
I've tried with detach and no_grad, it consume less memory but I can't get any id transfer running for at least 40 epochs with or without detaching

**Comment by taotaonice** [comment]
But I can get better attribution performance with detach

**Comment by ocastan** [comment]
Whatever I do, Y is almost like Xt and Xs does not seem to have any influence

**Comment by gerysand** [comment]
> But I can get better attribution performance with detach

hello, I'm an staff of a private fund, and quite interest on this application. May I have your mail and discuss some information with you? My mail gerysand#gmail.com (replace # with @).

Look forward to hear from you.

**Comment by ocastan** [comment]
Have you achieved anything interesting with detach ? I let it run with detach for 62 epochs and I didn't achieve any success far from the performance of the weights you sent to me

**Comment by mhmddpkts** [comment]
Discriminator loss is needed for adversarial training. In this code, disc. loss calculated with Xs and Y. If We detach Y, how Disc. give feedback to generator ? I think, Y.detach() should removed.
